### PR TITLE
client: add support to use the new "download" API

### PR DESCRIPTION
**Pull request by mvo5**
Now that we have the new /v2/download API in the daemon. This
PR adds the coresponding client API.

Needs PR#7510 to be fully useful.

### CODE REVIEW

File: client/export_test.go [hunk 1]
```diff
@@ -49,3 +49,5 @@ func UnmarshalSnapshotAction(body io.Reader) (act snapshotAction, err error) {
 	err = json.NewDecoder(body).Decode(&act)
 	return
 }
+
+type DownloadData downloadData
```
Comment (chipaca) [inline client/export_test.go]: you want a *type alias*, not a new type. I.e.,
```suggestion
type DownloadData = downloadData
```

File: client/snap_op.go [hunk 2]
```diff
@@ -305,3 +305,59 @@ func sendSnapFile(snapPath string, snapFile *os.File, pw *io.PipeWriter, mw *mul
...
 	pw.Close()
 }
+
+type snapRevisionOptions struct {
+	Channel   string `json:"channel,omitempty"`
+	Revision  string `json:"revision,omitempty"`
+	CohortKey string `json:"cohort-key,omitempty"`
+}
+
+type downloadData struct {
```
Comment (chipaca) [inline client/snap_op.go]: the `data` there doesn't add any information. Maybe `downloadAction` or `downloadInstruction`?

And yes, I know `snap_op` has `actionData`. That should clearly be `snapAction`. We've got a lot of things that have an Action in them: `appInstruction`, `snapshotAction`, `debugAction`, `aliasAction`, `CohortAction`, `InterfaceAction`, `warningsAction`, `actionData`, and`multiActionData`. The last two are clearly bad names: you have no idea what they are *for*. The rest could use some more uniformity, but no biggie.

File: client/snap_op.go [hunk 3]
```diff
@@ -305,3 +305,59 @@ func sendSnapFile(snapPath string, snapFile *os.File, pw *io.PipeWriter, mw *mul
...
+func (client *Client) Download(name string, options *SnapOptions) (suggestedFileName string, r io.ReadCloser, err error) {
+	if options == nil {
+		options = &SnapOptions{}
+	}
+	dlOpts := snapRevisionOptions{
+		Channel:   options.Channel,
+		CohortKey: options.CohortKey,
+		Revision:  options.Revision,
+	}
+	action := downloadData{
```
Comment (chipaca) [inline client/snap_op.go]: yeah, `downloadAction` would be fine :-)

**Comment by mvo5** [comment]
This is now updated for the changes in https://github.com/snapcore/snapd/pull/7560

**Pull request merged**